components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/ScopeMgtDAO.java

```diff
@@ -0,0 +1,527 @@
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2ScopeServerException;
+import org.wso2.carbon.identity.oauth2.Oauth2ScopeConstants;
+import org.wso2.carbon.identity.oauth2.bean.Scope;
+import org.wso2.carbon.identity.oauth2.util.NamedPreparedStatement;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.Map;
+import java.util.HashMap;
+
+/**
+ * Data Access Layer functionality for Scope management. This includes storing, updating, deleting and retrieving scopes
+ */
+public class ScopeMgtDAO {
+    private static final Log log = LogFactory.getLog(ScopeMgtDAO.class);
+
+    /**
+     * Add a scope
+     *
+     * @param scope    Scope
+     * @param tenantID tenant ID
+     * @throws IdentityOAuth2ScopeServerException
+     */
+    public void addScope(Scope scope, int tenantID) throws IdentityOAuth2ScopeServerException {
+
+        Connection conn = null;
+        PreparedStatement ps = null;
+        ResultSet rs;
+
+        try {
+            conn = IdentityDatabaseUtil.getDBConnection();
+
+            String scopeIdField = "SCOPE_ID";
+            if (conn.getMetaData().getDriverName().contains("PostgreSQL")) {
+                scopeIdField = "scope_id";
+            }
+
+            if (scope != null) {
+                ps = conn.prepareStatement(SQLQueries.ADD_SCOPE, new String[]{scopeIdField});
+                ps.setString(1, scope.getName());
+                ps.setString(2, scope.getDescription());
+                ps.setInt(3, tenantID);
+                ps.execute();
+
+                rs = ps.getGeneratedKeys();
+
+                int scopeID = 0;
+                if (rs.next()) {
+                    scopeID = rs.getInt(1);
+                }
+                // some JDBC Drivers returns this in the result, some don't
+                if (scopeID == 0) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("JDBC Driver did not return the scope id, executing Select operation");
+                    }
+                    ps = conn.prepareStatement(SQLQueries.RETRIEVE_SCOPE_ID_BY_NAME, new String[]{scopeIdField});
+                    ps.setString(1, scope.getName());
+                    ps.setInt(2, tenantID);
+                    rs = ps.executeQuery();
+
+                    if (rs.next()) {
+                        scopeID = rs.getInt(1);
+                    }
+                }
+
+                for (String binding : scope.getBindings()) {
+                    if (StringUtils.isNotBlank(binding)) {
+                        ps = conn.prepareStatement(SQLQueries.ADD_SCOPE_BINDING, new String[]{scopeIdField});
```

Comment:
Use different prepare statements/result set objects since when closing only the statement assign in last will be closed